server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java

```diff
@@ -0,0 +1,265 @@
+    }
+
+    /**
+     * Determines an index's current state, with respect to whether its mappings can
+     * be updated.
+     *
+     * @param clusterState the cluster state to use when calculating the upgrade state
+     * @param descriptor information about the system index to check
+     * @return a value that indicates the index's state.
+     */
+    UpgradeStatus getUpgradeStatus(ClusterState clusterState, SystemIndexDescriptor descriptor) {
+        final State indexState = calculateIndexState(clusterState, descriptor);
+
+        final String indexDescription = "Index [" + descriptor.getPrimaryIndex() + "] (alias [" + descriptor.getAliasName() + "])";
+
+        // The messages below will be logged on every cluster state update, which is why even in the index closed / red
+        // cases, the log levels are DEBUG.
+
+        if (indexState.indexState == IndexMetadata.State.CLOSE) {
+            logger.debug(
+                "Index {} is closed. This is likely to prevent some features from functioning correctly", indexDescription);
+            return UpgradeStatus.CLOSED;
+        }
+
+        if (indexState.indexHealth == ClusterHealthStatus.RED) {
+            logger.debug("Index {} health status is RED, any pending mapping upgrades will wait until this changes", indexDescription);
+            return UpgradeStatus.UNHEALTHY;
+        }
+
+        if (indexState.isIndexUpToDate == false) {
+            logger.debug(
+                "Index {} is not on the current version. Features relying "
+                    + "on the index will not be available until the index is upgraded",
+                indexDescription
+            );
+            return UpgradeStatus.NEEDS_UPGRADE;
+        } else if (indexState.mappingUpToDate) {
+            logger.trace("Index {} is up-to-date, no action required", indexDescription);
+            return UpgradeStatus.UP_TO_DATE;
+        } else {
+            logger.info("Index {} mappings are not up-to-date and will be updated", indexDescription);
+            return UpgradeStatus.NEEDS_MAPPINGS_UPDATE;
+        }
+    }
+
+    /**
+     * Updates the mappings for a system index
+     * @param descriptor information about the system index
+     */
+    private void upgradeIndexMetadata(SystemIndexDescriptor descriptor) {
+        if (isUpgradeInProgress.compareAndSet(false, true) == false) {
+            return;
+        }
+
+        final String indexName = descriptor.getPrimaryIndex();
+
+        PutMappingRequest request = new PutMappingRequest(indexName).source(descriptor.getMappings(), XContentType.JSON);
+
+        final OriginSettingClient originSettingClient = new OriginSettingClient(this.client, descriptor.getOrigin());
+
+        originSettingClient.admin().indices().putMapping(request, new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                isUpgradeInProgress.set(false);
```

Comment:
this sets to false early. There can still be in flight items and isUpgradeInProgress would be incorrect. You may want to use something like a `GroupedActionListener`

Reply:
I'm not sure I follow. The `upgradeIndexMetadata` method is called once for each index that needs upgraded, and the first thing the method does is attempt to set `isUpgradeInProgress` to `true`. So only one index will be upgraded, and the rest need to wait until `isUpgradeInProgress` becomes false again. I wrote it that way so as to be cautious about adding load to the master.

Were you thinking that we'd dispatch all the mapping updates in one go, and use a `GroupedActionListener` to release the `isUpgradeInProgress` once they'd all finished?

Reply:
> I'm not sure I follow. 

That's because I did not follow the code properly. I think it would be best to fire off all of the upgrades at once rather than sequentially since IIRC the code is computing everything that needs upgraded each time.

> Were you thinking that we'd dispatch all the mapping updates in one go, and use a GroupedActionListener to release the isUpgradeInProgress once they'd all finished?

Yes that is what I was thinking. I wanted to avoid resending each one every time there was a cluster state update if other cluster state updates happened prior to the mapping updates that we had previously sent.

Reply:
I've reworked the locking and upgrading to use a `GroupedActionListener`.